### DISCUSSION TITLE
call-func with key-presses

### DISCUSSION THREAD

**Discussion by shara19** [jspsych/jsPsych#873] [Q&A]
Hi,

My ultimate objective is to
1. load a random game 
2. record key-presses and response-times while the subject is playing

Now, my current setup has these games(in js) in separate files. I'm trying to understand what's the cleanest way to meet my objectives. 

I considered using call-function as follows, but I'm not sure how to capture key_presses.  
 ```
// --------------------- RENDER A RANDOM GAME ---------------------
            var render = function() {
                var games_list = ['gvgai_avoidgeorge', 'gvgai_aliens']
                var game_name = jsPsych.randomization.sampleWithReplacement(games_list, 1)[0];
                var subject_id = jsPsych.randomization.randomID(10);
                window.location.href = `games/${game_name}/0.html`;
                jsPsych.data.addProperties({subject: subject_id, game: game_name});
            };

// --------------------- TRIAL INITIALIZATION ---------------------
            var trial = {
            type: 'call-function',
            func: render,


```
I tried considering using html-keyboard-response but the 'stimulus' seems to handle simple html components. Am I wrong?

Thanks!

**Comment by vijaymarupudi** [comment]
You cannot change `location.href` and keep the jsPsych data, when a new page is loaded, the JavaScript context of the current page is lost.

You might want to either use an iframe, a new window (see the `window.postMessage()` function), or look into the localStorage API.

~ Vijay

**Reply by shara19** [reply]
Thank you. This was helpful.

I did create an iframe to update jsPsych.data object. However, I'm struggling to send the jsPsych.data object to another module which processes the game and should update the data with key-presses array. Do you have any reference documentation to something like this?

**Comment by shara19** [comment]
My current setup is
Main:
-jspsych-6.10
-games
--play_game.js
--play_game.html
-index.html
-index.js

I'm loading jspsych in the index.html and using index.js to render as below:
```

// --------------------- INSTRUCTIONS PAGE ---------------------
var start_page = {
    type: 'html-button-response',
    stimulus: '<p>Hello! Ready to play?</p>',
    choices: ['Play'],
};

// --------------------- RENDER A RANDOM GAME ---------------------
var render = function() {

    var games_list = ['gvgai_avoidgeorge'] // TODO: extend list
    var game_name = jsPsych.randomization.sampleWithReplacement(games_list, 1)[0];
    var subject_id = jsPsych.randomization.randomID(10);

    var ifrm = document.createElement('iframe');
    ifrm.setAttribute('id', 'ifrm');
    ifrm.src = `games/${game_name}/0.html`;
    document.body.appendChild(ifrm);

    jsPsych.data.addProperties({subject: subject_id, game: game_name});
};


// function set_action (actionList){
//     console.log('set action')
//     jsPsych.data.addProperties({'action': actionList});
// }

// --------------------- TRIAL INITIALIZATION ---------------------
var trial = {
type: 'call-function',
func: render,
};

// --------------------- INIT ---------------------
jsPsych.init({
    timeline: [start_page, trial]

})


```

I'm trying to import index.js to play_game.js to update, but says jsPsych isn't initialized. When I try to import jspsych in play_game.js, I think it creates a new object altogether. Is there a way to retain (and update) the object? I'm very new to js, I'm pretty comfortable with python (if that helps)

**Reply by shara19** [reply]
Update:

I was able to solve the problem using window.parent.jsPsych object within in play_game.js. and I added this bit to init
```
on_finish: function() {
            jsPsych.data.get().localSave('csv','mydata.csv');}
```

However, the current problem I'm having is that it is saving the data at the beginning instead of after-trial completion because the iframe is getting loaded independently. Any recommendations on how to circumvent this?

**Reply by vijaymarupudi** [reply]
You should not manipular `document.body` for this use case. Try having a look at the code of this plugin: https://www.jspsych.org/plugins/jspsych-external-html for some hints.

Use [`jsPsych.getDisplayElement()`](https://www.jspsych.org/core_library/jspsych-core/#jspsychgetdisplayelement) instead.

You're approaching this the wrong way. Your game html page should have no awareness of jsPsych. It should just start the game and after it's done, should send a message to the jsPsych page using `window.parent.postMessage()`. In the `func:` property function, you should listen for that message, and then do all the necessary jsPsych things, like adding properties and so on.

As for the problem with jsPsych saving data at the beginning instead of waiting, you'll need the [`async`](https://www.jspsych.org/plugins/jspsych-call-function/#async-function-call) parameter.
